pytasa/fundamental.py

```diff
@@ -8,7 +8,407 @@
+    VR = np.dot(V,RR) 
+    return VR
+    
+    
+def _rayvel_old(C,SN,rho):
+    """
+    TO CALCULATE THE RAY-VELOCITY VECTOR CORRESPONDING TO A SLOWNESS VECTOR.
+    Original fortran by David Mainprice as part of the EMATRIX code.
+    Converted to Python by Alan Baird
+    
+    C: Stiffness tensor in Voigt Notation (6X6).
+    SN: Slowness vector (3).
+    rho: Density
+    
+    returns VG: Group velocity vector (3)
+    
+    
+    """
+    
+    ijkl = np.array([[0,5,4],
+                     [5,1,3],
+                     [4,3,2]])
+
+    
+    gamma = np.array([[ SN[0],   0.0,   0.0,   0.0, SN[2], SN[1]],
+                      [   0.0, SN[1],   0.0, SN[2],   0.0, SN[0]],
+                      [   0.0,   0.0, SN[2], SN[1], SN[0],  0.0]])
+    F = np.dot(np.dot(gamma,C),np.transpose(gamma))-np.identity(3)*rho
+    
+    
+    
+    # Signed cofactors of F[i,k]
+    CF = np.zeros((3,3))
+    
+    CF[0,0]=F[1,1]*F[2,2]-F[1,2]**2
+    CF[1,1]=F[0,0]*F[2,2]-F[0,2]**2
+    CF[2,2]=F[0,0]*F[1,1]-F[0,1]**2
+    CF[0,1]=F[1,2]*F[2,0]-F[1,0]*F[2,2]
+    CF[1,0]=CF[0,1]
+    CF[1,2]=F[2,0]*F[0,1]-F[2,1]*F[0,0]
+    CF[2,1]=CF[1,2]
+    CF[2,0]=F[0,1]*F[1,2]-F[0,2]*F[1,1]
+    CF[0,2]=CF[2,0]
+    
+    # Derivatives of determinant elements
+    DF = np.zeros((3,3,3))
+    for i in range(3):
+        for j in range(3):
+            for k in range(3):
+                DF[i,j,k]=0.0
+                for l in range(3):
+                    
+                    DF[i,j,k] = DF[i,j,k] + (C[ijkl[i,j],ijkl[k,l]] + C[ijkl[k,j],ijkl[i,l]]) * SN[l]
+                    
+    # Components of gradient
+    DFD = np.zeros(3)
+    for k in range(3):
+        DFD[k]=0.0
+        for i in range(3):
+            for j in range(3):
+                DFD[k]=DFD[k]+DF[i,j,k]*CF[i,j]
+                
+    # Normalize to obtain group velocity
+    denom = 0.0
+    VG = np.zeros(3)
+    for i in range(3):
+        denom=denom+SN[i]*DFD[i]
+    for i in range(3):
+        VG[i]=DFD[i]/denom
+                
+            
+    return VG
+    
+def _rayvel(Cin,SN,rho):
+    """
+    Calculate the ray-velocity vector corresponding to a slowness vector.
+    Based on original fortran code by Mike Kendall and Sean Guest as part of ATRAK.
+    Converted to Python by Alan Baird
+    
+    C: Stiffness tensor in Voigt Notation (6X6).
+    SN: Slowness vector (3).
+    rho: Density
+    
+    returns VG: Group velocity vector (3)
+    
+    
+    """
+    
+    C = Cin/rho
+    
+    
+    ijkl = np.array([[0,5,4],
+                     [5,1,3],
+                     [4,3,2]])
+    
+    
+
+    
+    gamma = np.array([[ SN[0],   0.0,   0.0,   0.0, SN[2], SN[1]],
+                      [   0.0, SN[1],   0.0, SN[2],   0.0, SN[0]],
+                      [   0.0,   0.0, SN[2], SN[1], SN[0],  0.0]])
+    cp = np.dot(np.dot(gamma,C),np.transpose(gamma))
+    
+    dcp = np.zeros((3,3,3))
+    for r in range(3):
```

Comment:
I suspect this can be rewritten without the loops (in the same way that `_velo` was modified. This ought to be way quicker, but lets leave this for now and get something that works and has tests before we try to optimise.